pkg/scheduler/watchers/membercluster/controller.go

```diff
@@ -0,0 +1,222 @@
+	// that this controller will not emit false negatives, i.e., all the changes that require
+	// the scheduler's attention will be captured; in other words, false positives may still
+	// happen, i.e., this controller may trigger the scheduler to run a scheduling loop even though
+	// there is no change needed. In such situations, the scheduler will be the final judge and
+	// process the CRPs accordingly.
+
+	// Retrieve the member cluster.
+	memberCluster := &fleetv1beta1.MemberCluster{}
+	memberClusterKey := types.NamespacedName{Name: req.Name}
+	isMemberClusterMissing := false
+
+	memberClusterGetErr := r.Client.Get(ctx, memberClusterKey, memberCluster)
+	switch {
+	case errors.IsNotFound(memberClusterGetErr):
+		// On very unlikely occasions, it could happen that the member cluster is deleted
+		// before this controller gets a chance to process it, it may happen when a member cluster
+		// leaves the fleet. In such cases, this controller will request the scheduler to check
+		// all CRPs just in case.
+		isMemberClusterMissing = true
+	case memberClusterGetErr != nil:
+		klog.ErrorS(memberClusterGetErr, "Failed to get member cluster", "memberCluster", memberClusterRef)
+		return ctrl.Result{}, controller.NewAPIServerError(true, memberClusterGetErr)
+		// Do nothing if there is no error returned.
+	}
+
+	// List all CRPs.
+	//
+	// Note that this controller reads CRPs from the same cache as the scheduler.
+	crpList := &fleetv1beta1.ClusterResourcePlacementList{}
+	if err := r.Client.List(ctx, crpList); err != nil {
+		klog.ErrorS(err, "Failed to list CRPs", "memberCluster", memberClusterRef)
+		return ctrl.Result{}, controller.NewAPIServerError(true, err)
+	}
+
+	crps := crpList.Items
+	if !isMemberClusterMissing && memberCluster.Spec.State == fleetv1beta1.ClusterStateJoin {
+		// If the member cluster is set to the left state, the scheduler needs to process all
+		// CRPs (case 2c)); otherwise, only CRPs of the PickAll type + CRPs of the PickN type,
+		// which have not been fully scheduled, need to be processed (case 1a) and 1b)).
+		crps = classifyCRPs(crpList.Items)
+	}
+
+	// Enqueue the CRPs.
+	//
+	// Note that all the CRPs in the system are enqueued; technically speaking, for situation
+	// 1a) and 1b), PickN CRPs that have been fully scheduled needs no further processing, however,
+	// for simplicity reasons, this controller will not distinguish between the cases.
+	for idx := range crps {
+		crp := &crps[idx]
+		klog.V(2).InfoS(
+			"Enqueueing CRP for scheduler processing",
+			"memberCluster", memberClusterRef,
+			"clusterResourcePlacement", klog.KObj(crp))
+		r.SchedulerWorkQueue.AddRateLimited(queue.ClusterResourcePlacementKey(crp.Name))
+	}
+
+	// The reconciliation loop completes.
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager builds a controller with Reconciler and sets it up with a controller manager.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	customPredicate := predicate.Funcs{
```

Comment:
This function has no logs at all. We will be in the dark if anything unexpected happens.